File: Raspberry Pi/GPIO Examples/cpp/gpio_example.cpp
/**
 ******************************************************************************
 * @file    /gpio_examples/cpp/gpio_example.cpp
 * @author  Adrian Wojcik
 * @version V1.0
 * @date    15-Mar-2020
 * @brief   Raspberry Pi GPIO control: C++ with wiringPi lib
 ******************************************************************************
 */

#include <iostream>
#include <future>
#include <thread>
#include <chrono>
#include <wiringPi.h>

int main()
{
	const int led = 28; //< Red LED: Physical pin 38, BCM GPIO20, and WiringPi pin 28.
	const int button = 27; //< Push-button: Physical pin 36, BCM GPIO16, and WiringPi pin 27.
	
	bool btn_state = true, btn_state_last = true;
	bool led_state = false;
	
	std::chrono::milliseconds timeout(100);
	std::future<int> async_getchar = std::async(std::getchar);

	wiringPiSetup();

	pinMode(button, INPUT);
	pinMode(led, OUTPUT);
	
	digitalWrite(led, LOW);
	
	std::cout << "Press ENTER to exit." << std::endl;
	
	while(1) 
	{
		btn_state = (digitalRead(button) == HIGH);

		if( !btn_state && btn_state_last){
			led_state = !led_state;
			digitalWrite(led, led_state);
			std::cout << "LED state: ";
			if(led_state)			
				std::cout<< "ON" << std::endl;
			else
				std::cout << "OFF" << std::endl;
		}

		btn_state_last = btn_state;

		if(async_getchar.wait_for(timeout) == std::future_status::ready)
		{
			async_getchar.get();	
			break;
		}			
	}
    
	return 0;
}

File: Raspberry Pi/GPIO Examples/python/gpio_example.py
#!/usr/bin/python3
#**
#******************************************************************************
#* @file    /gpio_examples/python/gpio_example.py
#* @author  Adrian Wojcik
#* @version V1.0
#* @date    15-Mar-2020
#* @brief   Raspberry Pi GPIO control: Python 3 with RPi.GPIO lib
#******************************************************************************

import time
import sys
import select
try:
	import RPi.GPIO as GPIO
except RuntimeError:
	print("Error importing RPi.GPIO! This is probably because you need superuser privileges. You can achieve this by using 'sudo' to run your script")
	
timeout = 0.1
buttonState = True
buttonStateLast = True
ledState = False
ledStateName = ["OFF", "ON"]
i = ''

# Pin Definitons:
ledPin = 38    #< LED: Physical pin 38, BCM GPIO28
buttonPin = 36 #< Push-button: Physical pin 36, BCM GPIO16

# Pin Setup:
GPIO.setmode(GPIO.BOARD)
GPIO.setup(ledPin, GPIO.OUT) # LED pin set as output
GPIO.setup(buttonPin, GPIO.IN) # Button pin set as input 

GPIO.output(ledPin, ledState)

print("Press ENTER to exit.")

while not i:
	# Waiting for I/O completion
	i, o, e = select.select( [sys.stdin], [], [], timeout )

	if (i):
		sys.stdin.readline();
		GPIO.cleanup() # cleanup all GPIO
		exit()

	buttonState = (GPIO.input(buttonPin) == GPIO.HIGH)
	
	if (buttonState is False) and (buttonStateLast is True):
		ledState = not ledState
		GPIO.output(ledPin, ledState)
		print("LED state:",ledStateName[ledState])
		
	buttonStateLast = buttonState
